### PR TITLE
Signup: Trying siteless checkout

### PR DESCRIPTION
**Pull request by mmtr**
#### Changes proposed in this Pull Request

This PR explores siteless checkouts as a way of avoiding duplicate sites during signup. 

Current code is only a POC that modifies the default onboarding process so sites are not created when completing the last step (select a plan). 

At this point, the only remaining work is to make sure the transactions API endpoint creates the site once the payment has been processed.

Any assistance here would be highly appreciated.

Further reading: pbAPfg-5U-p2
Part of #39424

#### Testing instructions

* Apply D39598-code and sandbox the API.
* Go to `/start`.
* Complete all the steps and select a paid plan.
* Make sure you're redirected to the checkout page and the site has not been created.
* Make sure the checkout displays a shopping cart that contains the selected pain plan.
* **Do not proceed with the payment**. D39598-code still doesn't create sites after the plan is purchased, so you might end up adding the plan to a wrong site.
* Make sure the signup flow remains unmodified (sites created before checkout) when turning off the `signup/siteless-checkout` feature flag.

### CODE REVIEW

**Comment by sirbrillig** [comment]
Hm. I don't know a lot about signup, but it makes some sense that something is expecting to query domains for a site that doesn't exist. There's already flows that allow signup without a site; maybe it's worth examining one of those to see how they work? For example, visiting `/domains` and then selecting "Just buy a domain".

File: client/signup/main.jsx [hunk 1]
```diff
@@ -525,6 +525,7 @@ class Signup extends React.Component {
 		const isDomainsForSiteEmpty =
 			this.props.isLoggedIn &&
 			this.props.signupDependencies.siteSlug &&
+			this.props.signupDependencies.siteSlug !== 'no-site' &&
```
Comment (sirbrillig) [inline client/signup/main.jsx]: 👍

Comment (mmtr) [reply client/signup/main.jsx]: ❤️ Thanks @kwight! This solves the JS error I was seeing. The checkout page is still not loaded, so I guess the signup code is still waiting for a new site to be created before doing the redirect. Will keep investigating to see if I can identify where.

**Comment by mmtr** [comment]
> there is something that is preventing the checkout page from showing up (the page gets stuck in the "Hooray! Your site will be ready shortly" screen).

For some reason, the `domains-with-preview` step is not marked as completed, so the signup flow controller does not execute the `onComplete` callback:

https://github.com/Automattic/wp-calypso/blob/05b683bfe04ff4144ac49e1896e9f8461fcdf718/client/lib/signup/flow-controller.ts#L257-L262

I checked the domain-only flow and I think we're replicating the same steps:
- Both are [submitting the data and moving to the next step](https://github.com/Automattic/wp-calypso/blob/05b683bfe04ff4144ac49e1896e9f8461fcdf718/client/signup/steps/domains/index.jsx#L284-L300).
- Both are [creating a cart](https://github.com/Automattic/wp-calypso/blob/05b683bfe04ff4144ac49e1896e9f8461fcdf718/client/lib/signup/step-actions.js#L69-L80) without creating a site.

I still have not found what's keeping the step uncompleted.

**Comment by mmtr** [comment]
> I still have not found what's keeping the step uncompleted.

Found it! The issue was actually in the last step (pick a plan) which calls `processItemCart` and waits for the new site to be fetched:

https://github.com/Automattic/wp-calypso/blob/a06b1c8ccc77b53cf623c6a39f4e7fd82712fe14/client/lib/signup/step-actions.js#L352

I'm preparing a commit that updates `processItemCart` so it can handle siteless cart items.

**Comment by mmtr** [comment]
Minor update: D39598-code solves the empty product list on the API but the signup still freezes on the "we're creating your site" screen.

Some debugging indicates this is caused by the state layer:
- After selecting a paid plan, we dispatch a [`completeSignupStep`](https://github.com/Automattic/wp-calypso/blob/1cba443d9274acfd18a1b9bde3daccc357ecc9f1/client/state/signup/progress/actions.js#L106) action that marks the `domains-siteless` step as completed.
- Immediately after, we dispatch somewhere a [`processStep`](https://github.com/Automattic/wp-calypso/blob/1cba443d9274acfd18a1b9bde3daccc357ecc9f1/client/state/signup/progress/actions.js#L115) action for the same `domain-siteless` step that overrides the previous completion and marks the step as processing.
- We then dispatch another `completeSignupStep` action for marking the `plans-siteless` step as completed.
- There are no more actions, but seems `domains-siteless` remains processing, the signup does not trigger the `onComplete` callback.

Is anyone familiar with the signup state layer who could assist with this and find out why the `domains-siteless` step is being overridden?

**Comment by mmtr** [comment]
Almost there! f67bc6e fixed the step override I was experiencing before and signup flow finally redirects to the checkout page 🎉 

We have now a new fresh error there, but hopefully it'll be easier to work around:

Screen Shot 2020-02-28 at 17 28 42

**Comment by kwight** [comment]
I was able to get a working checkout form displayed, by bypassing the importance of a `selectedSite` in 08ea786. This allows the call to `/me/transactions` to be made.

It gets fun here, though. At first I was getting a failed error of "This blog already has a plan which supersedes the one you're trying to purchase.". It seems this was because the pseudo-site for siteless domains already had a plan for whatever reason (although I couldn't confirm this in internal tools, because of MC issues), so me trying to apply a personal plan in testing would leave me hanging at checkout. Once I forced my way through that failing check, I then ran into the deep internals of billing, and got lost trying to see where would be the best place to defeat the requirement of applying a plan to a site when purchasing a plan (billing _really_ wants the purchased plan to be applied to a site). I was seeing failures in `payment_pre_process` during `$checkout->process_transaction()`.

On checkout, we do know that we've come from signup because of the `?signup=1` param, so I'm thinking if we can get that value deeper into billing, during its various steps to process the transaction, a check for signup and the blog ID of the siteless domains site could allow the process to complete successfully. While we can now get `/me/transactions` to complete, it'll still return with `failed_purchases` values in the response (which we'd have to reconcile somewhere too).

**Comment by klimeryk** [comment]
> It gets fun here, though. At first I was getting a failed error of "This blog already has a plan which supersedes the one you're trying to purchase.". It seems this was because the pseudo-site for siteless domains already had a plan for whatever reason

What is happening is the `pre_purchase` check is still operating on `blog_id = 0`. And apparently we have three plans assigned to `blog_id = 0`... Two look like they are from these recent tests.

It should be fairly simple to either add checks for `empty` blog_id - looks like they are a few places. It will also speed up things nicely :)

I'd strongly recommend pairing with someone from Neutron on this - if you're getting into the billing code it will go much faster if you're working with someone experienced with it. @rralian @gwwar could you make it happen? 🙏

**Comment by mmtr** [comment]
> got lost trying to see where would be the best place to defeat the requirement of applying a plan to a site when purchasing a plan (billing _really_ wants the purchased plan to be applied to a site). I was seeing failures in `payment_pre_process` during `$checkout->process_transaction()`.

I think that requirement is part of the same `pre_purchase` function mentioned by @klimeryk, with a different though: duplicate plan checks are in the specific bundles products (personal, premium, etc) while the generic interface for bundles still relies on the blog id for some other checks (i.e. "Unable to buy <PLAN_NAME>, this site has <PLAN_NAME> purchased by different user.").

**Comment by kwight** [comment]
We'll need to keep an eye on pau2Xa-Vs-p2.

**Comment by gwwar** [comment]
> I'd strongly recommend pairing with someone from Neutron on this - if you're getting into the billing code it will go much faster if you're working with someone experienced with it

@klimeryk , @kwight is going to write up a recap post for context and we'll chat with folks

**Comment by kwight** [comment]
> @kwight is going to write up a recap post for context and we'll chat with folks

pbAPfg-7C-p2

**Comment by millerf** [comment]
> pbAPfg-7C-p2

@kwight What are those codes? Is it only internal?

> Apply D39598-code and sandbox the API.

@mmtr How does that work and/or it is possible for 3rd parties to help out?

**Comment by mmtr** [comment]
Hi @millerf! Yeah, both of them refer to internal tools. 
- pbAPfg-7C-p2 represents an internal post written by @kwight where we further discuss our findings with other colleagues who are more familiar with the codebase we're modifying. We try to share here a summary of that discussion, so folks without access to it can be aware. We forgot to do that here (mainly because the discussion is still happening), so apologies for that! Our major concern so far is that the required billing changes for this PR to work will affect some marketing email campaigns (they rely on sites being created before checkout), but we're still leaning to try this and see how this flow impacts conversion.
- D39598-code represents an internal patch with the required billing changes in the API for this PR to work. Since that code is not public, external contributors cannot test this until that patch is deployed to production. I'll update the summary once this PR doesn't depend on any internal patch.

**Comment by millerf** [comment]
@mmtr Thanks for the clearing those things out! 

>I'll update the summary once this PR doesn't depend on any internal patch.
I'll keep an eye out on this thread, and see if I can help out once the patch is published.

Thanks!

**Comment by mmtr** [comment]
This was already implemented in a different way. See #45040

**Pull request closed without merging**